exercises/hardware_access_control/README.md

```diff
@@ -0,0 +1,160 @@
+The keen eyed among you will have noticed that this is already happening with both toggling user LED 7.
+
+## Part 2
+
+Let's introduce some access control for the LEDs.
+To do this, we can create a new compartment `gpio_access` with sole access to the GPIO MMIO region.
+This compartment will arbitrate access to the LED outputs by making use of CHERIoT's sealing mechanism.
+When a compartment seals a capability, it can no longer be dereferenced or modified until it is unsealed by a compartment with the capability to do so.
+`gpio_access` uses these sealing capabilities as LED *handles* that it can give to other compartments.
+These other compartments can't use the handles directly, but can only pass them to `gpio_access` which can unseal them and use them.
+In this case, they only point to a `LedHandle` structure that only holds the index of a LED.
+They are purely used as a proof of LED ownership.
+*For more information on sealing, see the [`cheriot-rtos/examples/05.sealing/`][].*
+
+[`cheriot-rtos/examples/05.sealing/`]: ../../cheriot-rtos/examples/05.sealing/
+
+`blinky_raw` and `led_walk_raw` have been adapted to use this new compartment and renamed `blinky_dynamic` and `led_walk_dynamic`.
+You'll notice these compartments use `add_deps` in the [`xmake.lua`][] file to declare that they depend on `gpio_access`.
+Take a moment to look at the sources for these compartments in [`exercises/hardware_access_control/part_2/`][].
+
+[`exercises/hardware_access_control/part_2/`]: ../../exercises/hardware_access_control/part_2/
+
+If you now run the `hardware_access_part_2` firmware on the FPGA, you'll notice only `blinky_dynamic` is toggling it's LED.
+Looking at the [UART console from the FPGA][running on fpga], the following message will pop up.
+
+```
+hardware_access_control/part_2/led_walk_dynamic.cc:19 Assertion failure in start_walking
+LED 0x7 couldn't be aquired
+```
+
+`led_walk_dynamic` was run after `blinky_dynamic` because it's thread was given a lower priority in the [`xmake.lua`][].
+So when it asked for access to user LED 7, it was denied by `gpio_access` because this LED had already been given to `blinky_dynamic`.
+
+Now change `NumLeds` in [`led_walk_dynamic.cc`][] from 8 to 7, then rebuild.
+Both compartments should run happily again.
+Not only will both compartments run happily, but `led_walk_dynamic` will output the following over the console.
+
+[`led_walk_dynamic.cc`]: ../../exercises/hardware_access_control/part_2/led_walk_dynamic.cc
+
+```
+Led Walk Dynamic:           LED 3 Handle: 0x1087d0 (v:1 0x1087d0-0x1087e0 l:0x10 o:0xc p: G RWcgm- -- ---)
+Led Walk Dynamic: Destroyed LED 3 Handle: 0x1087d0 (v:1 0x1087d0-0x1087e0 l:0x10 o:0xc p: G RWcgm- -- ---)
+Led Walk Dynamic:       New LED 3 Handle: 0x108878 (v:1 0x108878-0x108888 l:0x10 o:0xc p: G RWcgm- -- ---)
+```
+
+These come from some superfluous lines in [`led_walk_dynamic.cc`][], that releases ownership of user LED 3 only to later reacquire ownership.
```

Comment:
> These come from superfluous lines in `led_walk_dynamic.cc`, that releases ownership of user LED 3 ...`

Grammar: These come from superfluous lines in `led_walk_dynamic.cc`, **which release** ownership of user LED 3 ...

("lines" is plural, "that releases" is third-person singular)

Reply:
Addressed in: https://github.com/lowRISC/sonata-software/pull/42